cgrok/tree.c

```diff
@@ -0,0 +1,198 @@
+#include "tree.h"
+#include <stdint.h>
+#include <string.h>
+
+// A shim to use dictlib trees instead of TC
+struct TCTREE {
+  dict *dict;
+  dict_itor *iter;
+};
+
+// Toyko cabinet accepts any arbitrary-sized data as a tree key.
+// We support this by using the first 4 bytes of the key to store the length.
+// So we have to wrap the libdict comparators with implementations that skip the first four bytes
+// of the key
+int tccmpint32(const void* k1, const void* k2) {
+  return dict_int_cmp(k1+4, k2+4);
+}
+
+int dict_var_str_cmp(const void* k1, const void* k2) {
+  return dict_str_cmp(k1+4, k2+4);
+}
+
+void tcfree(void *key, void *value) {
+  free(key);
+  free(value);
+}
+
+TCTREE *tctreenew(void) {
+  return tctreenew2(dict_var_str_cmp, 0);
+}
+
+TCTREE *tctreenew2(dict_compare_func cp, void *cmpop) {
+  TCTREE *tree = malloc(sizeof(TCTREE));
+  tree->dict = hb_dict_new(cp, tcfree);
+  tree->iter = NULL;
+  return tree; 
+}
+
+// Iterate over elements in ascending key order
+void tctreeiterinit(TCTREE *tree) {
+  if (tree->iter != NULL) {
```

Comment:
I guess tokoyo cabinet isn't threadsafe?  This would be pretty easy for one to iterate 1/2 way through the dict, then someone else starts iterating and then the first person returns to further iterate only to find their iteration point has been reset?